enclave-runner/src/usercalls/mod.rs

```diff
@@ -545,6 +565,37 @@ fn trap_attached_debugger(tcs: usize) {
     }
 }
 
+/// UsercallExtension provides a mechanism for the enclave code interface with an external service via a modified runner.
+/// If the implemenentation of the functions of UsercallExtension reutrn None, then [default functions](../../fortanix_sgx_abi/struct.Usercalls.html#networking)
+/// would be called which inteface the enclave with TCP stream.
+/// An implementation of UsercallExtension, can be registered while [building](../struct.EnclaveBuilder.html#method.usercall_ext) the enclave.
+
+#[allow(unused)]
+pub trait UsercallExtension : 'static + Send + Sync + std::fmt::Debug {
+    /// Allows the enclave to connect to an external service specified by the addr bytes.
+    /// The Runner should determine the service that the enclave is trying to connect to by parsing the addr bytes.
+    /// If `connect_stream` returns None, the default implementation of [connect_stream](../../fortanix_sgx_abi/struct.Usercalls.html#method.connect_stream) would be called.
+    /// Apart from connecting to the external service, the behavior of this function should be same as the default implementation.
+    fn connect_stream(
+        &self,
+        addr: &[u8],
+        local_addr: Option<&mut String>,
+        peer_addr: Option<&mut String>,
+    ) -> IoResult<Option<Box<SyncStream>>> {
+        Ok(None)
```

Comment:
I had earlier thought, when we add more functions here like bind and accept, user could implement a subset of them, hence I kept a default. 
I am open if we keep the policy that the user must implement a function for all members, which could, in some cases, just return None.
Please suggest which way to proceed.

Reply:
Ok, if/when we add custom usercalls for `bind` and `accept`, I could see making this method optional. Until then, I think it should be required. Note that it is backward compatible to make a required method optional, but not to make an optional method required.

Reply:
The default impl is necessary when this trait is extended as @VardhanThigle explained. I want to review the implementation now instead of in a follow-up PR that otherwise doesn't touch the connect logic.

Reply:
Sure. Reverted back the default implementation. Will start working on the accept and bind once this gets merged.